plugins/task-plugin/src/che-workspace-client.ts

```diff
@@ -105,34 +44,16 @@ export class CheWorkspaceClient {
   }
 
   async getMachineExecServerURL(): Promise<string> {
-    const machineExecServer = await this.getMachineExecServer();
-    if (!machineExecServer) {
-      throw new Error(`No server with type ${TERMINAL_SERVER_TYPE} found.`);
-    }
-    if (!machineExecServer.attributes!.port) {
-      throw new Error('No machine-exec-server attributes.port found');
-    }
-    return `ws://127.0.0.1:${machineExecServer.attributes!.port}`;
+    const machineExecEndpoint = await this.getMachineExecEndpoint();
+    const port = machineExecEndpoint.attributes?.port || machineExecEndpoint.attributes?.targetPort;
+    return `ws://127.0.0.1:${port}`;
   }
 
-  protected async getMachineExecServer(): Promise<cheApi.workspace.Server | undefined> {
-    const machines = await this.getMachines();
-    for (const machineName in machines) {
-      if (!machines.hasOwnProperty(machineName)) {
-        continue;
-      }
-      const servers = machines[machineName].servers!;
-      for (const serverName in servers) {
-        if (!servers.hasOwnProperty(serverName)) {
-          continue;
-        }
-
-        const serverAttributes = servers[serverName].attributes;
-        if (serverAttributes && serverAttributes['type'] === TERMINAL_SERVER_TYPE) {
-          return servers[serverName];
-        }
-      }
+  protected async getMachineExecEndpoint(): Promise<che.endpoint.ExposedEndpoint> {
+    const terminalEndpoints = await che.endpoint.getEndpointsByType(TERMINAL_SERVER_TYPE);
+    if (terminalEndpoints.length === 1) {
+      return terminalEndpoints[0];
     }
-    return undefined;
+    throw new Error(`Unable to find terminal with type ${TERMINAL_SERVER_TYPE}: Found: ${terminalEndpoints}`);
```

Comment:
`Unable to find terminal ...`
but `Found: ${terminalEndpoints}`
It looks odd to me )

Reply:
yes in fact terminalEndpoints are list of what is retrieved, it can be more than 1.

so do you suggest something like: `Unable to find a single endpoint with type ${TERMINAL_SERVER_TYPE}: Found ${terminalEndpoints}`

Reply:
It's a minor thing, but yeah... `a single` looks a bit clearer.

Reply:
Ok if it's the only feedback, I'll do a follow up PR as currently all 11 checks are green :-)